### PR TITLE
F m1 auth db

### PR DESCRIPTION
**Pull request by BanerjeeAgniva**
## Summary

Implements the Auth module database foundation using Flyway migrations, including schema creation, audit-safe role management tables, and seeded RBAC data for built-in roles and permissions.

---

## What Changed

* Added auth module dependencies for Spring Data JPA and Spring Security in `auth/pom.xml`
* Created Flyway migration structure under `app/src/main/resources/db/migration/`
* Added schema migrations for permissions, roles, role_permissions, users, api_keys, and role_audit_logs
* Added seed migrations for 39 permissions, 12 built-in roles, and all role-permission mappings from `ROLE-PERMISSIONS.md`
* Added indexes and FK constraints for optimized lookup/query performance
* Implemented append-only audit logging design with intentionally decoupled audit foreign keys

---

## Why This Change?

The auth module previously had no database implementation, preventing RBAC, user authentication, API key management, and audit tracking from functioning. This change establishes the foundational schema and seeded authorization model required for all downstream authentication and authorization features.

---

## Testing

* [x] Tested locally
* [x] Edge cases considered
* [x] No breaking changes introduced

### Test Evidence

```bash
$ ./mvnw flyway:migrate

Successfully applied 9 migrations to schema "public"

Schema version: 9
Current version: 9
State: Success
```

```sql
SELECT COUNT(*) FROM permissions;
-- 39

SELECT COUNT(*) FROM roles;
-- 12

SELECT COUNT(*) FROM role_permissions;
-- verified against ROLE-PERMISSIONS.md
```

```bash
Hibernate validation completed successfully.
Application started without schema errors.
```

image

image

image

---

### Areas Affected

* [] Backend
* [ ] Frontend
* [ ] Routing
* [ ] Infra
* [x] Database
* [ ] NA

### Modules Affected

* [x] M1 — Auth
* [ ] M2 — Pricing
* [ ] M3 — Grid
* [ ] M4 — Orders
* [ ] M5 — Dispatch
* [ ] M6 — Routing
* [ ] M7 — Hub
* [ ] M8 — Barcode
* [ ] M9 — Airline
* [ ] M10 — SLA
* [ ] M11 — Exceptions
* [ ] NA

---

## Risks / Concerns

* Seed migration consistency depends on `ROLE-PERMISSIONS.md` remaining aligned with future RBAC updates
* Migration ordering is critical due to FK dependencies between roles, permissions, users, and API keys
* Audit table intentionally omits FK constraints on actor/target references to preserve historical records after user deletion

---

## Checklist

* [x] Code follows project conventions (package layout, no cross-module internal imports)
* [x] Self-review completed
* [x] Append-only audit trail preserved (no mutations to scan/manifest/role-change records)
* [x] No sensitive data or credentials exposed
* [x] Documentation updated if behaviour changed
* [x] Ready for review

----

Read if you have time to waste (good time wasted hehehhehe)

---

## Chapter 13 — One Day in the Life of the Database

*A row-level walkthrough. Every action in Chapters 1–12 leaves a trace in the database. This is what that trace looks like.*

---

### 06:00 — Migrations Run, the Schema Exists, Seed Data Lands

The app starts for the first time. Flyway executes V1 through V9 in order.

After V7 completes, the `permissions` table holds 40 rows. A sample:

| id (UUID) | action |
|---|---|
| `a1b2c3...` | `shipment:create` |
| `d4e5f6...` | `shipment:view` |
| `g7h8i9...` | `hub:scan` |
| `j0k1l2...` | `api-key:create:own` |
| … 36 more … | … |

After V8, the `roles` table holds 12 rows:

| id (UUID) | name | display_name | city_scoped | is_builtin | active |
|---|---|---|---|---|---|
| `r1...` | `ADMIN` | Administrator | false | true | true |
| `r2...` | `STATION_MANAGER` | Station Manager | true | true | true |
| `r3...` | `DELIVERY_ASSOCIATE` | Delivery Associate | true | true | true |
| `r4...` | `B2B_USER` | B2B User | false | true | true |
| … 8 more … | … | … | … | … | … |

After V9, the `role_permissions` table holds 68 rows — one per (role, permission) pair. A sample for just `ADMIN`:

| role_id | permission_id |
|---|---|
| `r1...` | `a1b2c3...` ← shipment:create |
| `r1...` | `d4e5f6...` ← shipment:view |
| `r1...` | `g7h8i9...` ← hub:scan |
| … 9 more for ADMIN … | … |

No `users` rows yet. No `api_keys`. No `role_audit_logs`. The platform is an empty stage.

---

### 08:00 — Riya Logs In (No DB Write)

`POST /auth/login { "email": "admin@oneday.in", "password": "Admin1234!" }`

The `users` table is read (one SELECT by email, `active = true`). BCrypt comparison runs in memory. A JWT is issued. Nothing is written — login is a read-only operation on the DB.

---

### 09:15 — Riya Creates Arjun (First Row in `users`, First Row in `role_audit_logs`)

`POST /users { "email": "arjun.sharma@oneday.in", "role": "STATION_MANAGER", "cityId": "MUM" }`

Two DB writes happen:

**`users` table — 1 new row:**

| id | email | password_hash | name | role_id | city_id | active | must_change_password |
|---|---|---|---|---|---|---|---|
| `u-arjun` | arjun.sharma@oneday.in | `$2a$12$Kx...` | Arjun Sharma | `r2...` ← STATION_MANAGER | MUM | true | false |

`role_id` is a FK to `roles.id`. The bcrypt hash is 60 characters; the raw password `"Secure#9012"` is gone.

**`role_audit_logs` table — 1 new row:**

| id | actor_id | target_user_id | action | previous_role | new_role | city_id | reason |
|---|---|---|---|---|---|---|---|
| `log-1` | `u-riya` | `u-arjun` | CREATE | null | STATION_MANAGER | MUM | User registration |

`actor_id` and `target_user_id` are plain UUIDs — no FK to `users`. Even if Riya's account is later deleted, this row survives intact.

---

### 10:00 — Arjun's Staff Get Created (Three More `users` Rows)

Riya registers Priya (DELIVERY_ASSOCIATE / MUM), Rohan (HUB_OPERATOR / MUM), and Kavya (SUPERVISOR / MUM). Three rows land in `users`, three rows land in `role_audit_logs`.

**`users` after all three:**

| id | email | name | role_id (→ roles.name) | city_id | active |
|---|---|---|---|---|---|
| `u-arjun` | arjun.sharma@… | Arjun Sharma | STATION_MANAGER | MUM | true |
| `u-priya` | priya.desai@… | Priya Desai | DELIVERY_ASSOCIATE | MUM | true |
| `u-rohan` | rohan.mehta@… | Rohan Mehta | HUB_OPERATOR | MUM | true |
| `u-kavya` | kavya.nair@… | Kavya Nair | SUPERVISOR | MUM | true |

**`role_audit_logs` — 4 rows so far, all `action = CREATE`.**

---

### 11:30 — Farhan Signs Up (B2B User, No `city_id`)

`POST /users { "email": "farhan@farhankart.com", "role": "B2B_USER", "cityId": null }`

**`users` — new row:**

| id | email | name | role_id (→ roles.name) | city_id | active |
|---|---|---|---|---|---|
| `u-farhan` | farhan@farhankart.com | Farhan Khan | B2B_USER | **null** | true |

`B2B_USER` is not city-scoped. `city_id` is null — he can transact across all cities. One more `role_audit_logs` row with `action = CREATE`.

---

### 12:00 — Farhan Creates an API Key (First Row in `api_keys`)

`POST /auth/api-keys { "label": "farhankart-oms-prod" }`

The service generates a 32-byte random value → Base64-URL-encodes it → SHA-256-hashes it.

**`api_keys` table — 1 new row:**

| id | key_hash | user_id | label | active | last_used_at |
|---|---|---|---|---|---|
| `k-farhan-1` | `sha256:e3b0c4...` | `u-farhan` | farhankart-oms-prod | true | null |

The raw key `"AbCdEf12...XyZ"` is returned in the API response and never written anywhere. `last_used_at` is null — the key has not been used yet.

---

### 14:00 — FarhanKart's Backend Makes Its First Call (API Key Stamped)

FarhanKart sends `X-Api-Key: AbCdEf12...XyZ` on a shipment creation request. The filter hashes it, looks up the `api_keys` row, and stamps it.

**`api_keys` row updated:**

| id | key_hash | label | active | **last_used_at** |
|---|---|---|---|---|
| `k-farhan-1` | `sha256:e3b0c4...` | farhankart-oms-prod | true | **2026-05-13 14:00:03** |

This is the only mutable field on an api_key row. Everything else is write-once.

---

### 15:45 — Arjun Promotes Priya (Role Change, New `role_audit_logs` Row)

`PUT /users/u-priya/role { "newRole": "SUPERVISOR", "cityId": "MUM", "reason": "Strong performance" }`

**`users` row updated:**

| id | name | role_id (before) | role_id (after) | city_id |
|---|---|---|---|---|
| `u-priya` | Priya Desai | DELIVERY_ASSOCIATE → | **SUPERVISOR** | MUM |

The `role_id` FK value changes from the DELIVERY_ASSOCIATE UUID to the SUPERVISOR UUID.

**`role_audit_logs` — new row (row 6 total):**

| id | actor_id | target_user_id | action | previous_role | new_role | city_id | reason |
|---|---|---|---|---|---|---|---|
| `log-6` | `u-arjun` | `u-priya` | GRANT | DELIVERY_ASSOCIATE | SUPERVISOR | MUM | Strong performance |

The old row (`log-2`, `action = CREATE`, `new_role = DELIVERY_ASSOCIATE`) is untouched. The audit trail now tells a complete story: Priya was created as a DA, then promoted to SUPERVISOR by Arjun.

---

### 17:00 — The GHA Account Is Created (No `city_id` Enforcement)

`POST /users { "email": "gha.mum@indigo.in", "role": "AIRLINE_GHA", "cityId": "MUM" }`

| id | email | role_id (→ name) | city_id | active |
|---|---|---|---|---|
| `u-gha` | gha.mum@indigo.in | AIRLINE_GHA | MUM | true |

`AIRLINE_GHA` is not in the city-scoped list, so `city_id` was optional — Riya set it anyway for operational clarity. Their `role_permissions` entries cover only `manifest:view` and `handover:acknowledge`.

---

### 18:30 — A Rogue Operator Is Deactivated (Final State)

Rohan is found to have entered fraudulent scan events. Riya deactivates him:

`DELETE /users/u-rohan`

**`users` row updated:**

| id | name | role_id | active (before → after) |
|---|---|---|---|
| `u-rohan` | Rohan Mehta | HUB_OPERATOR | true → **false** |

**`role_audit_logs` — new row:**

| id | actor_id | target_user_id | action | previous_role | new_role | reason |
|---|---|---|---|---|---|---|
| `log-8` | `u-riya` | `u-rohan` | DEACTIVATE | HUB_OPERATOR | null | Account deactivated by admin |

Rohan's next HTTP request — even if his JWT hasn't expired — returns 403. The filter does a DB lookup, finds `active = false`, and refuses to set the SecurityContext.

---

### End of Day — Full Database Snapshot

**`permissions`** — 40 rows (static seed, unchanged)

**`roles`** — 12 rows (static seed, unchanged)

**`role_permissions`** — 68 rows (static seed, unchanged)

**`users`** — 6 rows:

| name | role | city_id | active |
|---|---|---|---|
| Riya (admin seed) | ADMIN | null | true |
| Arjun Sharma | STATION_MANAGER | MUM | true |
| Priya Desai | **SUPERVISOR** | MUM | true |
| Rohan Mehta | HUB_OPERATOR | MUM | **false** |
| Kavya Nair | SUPERVISOR | MUM | true |
| Farhan Khan | B2B_USER | null | true |
| GHA Mumbai | AIRLINE_GHA | MUM | true |

**`api_keys`** — 1 row (Farhan's `farhankart-oms-prod`, `active = true`, `last_used_at` stamped)

**`role_audit_logs`** — 8 rows, in order:

| # | actor | target | action | previous → new |
|---|---|---|---|---|
| 1 | Riya | Arjun | CREATE | — → STATION_MANAGER |
| 2 | Riya | Priya | CREATE | — → DELIVERY_ASSOCIATE |
| 3 | Riya | Rohan | CREATE | — → HUB_OPERATOR |
| 4 | Riya | Kavya | CREATE | — → SUPERVISOR |
| 5 | Riya | Farhan | CREATE | — → B2B_USER |
| 6 | Arjun | Priya | GRANT | DELIVERY_ASSOCIATE → SUPERVISOR |
| 7 | Riya | GHA | CREATE | — → AIRLINE_GHA |
| 8 | Riya | Rohan | DEACTIVATE | HUB_OPERATOR → — |

Every mutation to user identity — creation, promotion, deactivation — has a corresponding, permanent, FK-free row here. The table is append-only. No row in this table has ever been updated or deleted.

### CODE REVIEW

File: app/src/main/resources/db/migration/V5__create_api_keys.sql [hunk 1]
```diff
@@ -0,0 +1,13 @@
+CREATE TABLE api_keys (
+    id           UUID PRIMARY KEY DEFAULT gen_random_uuid(),
+    key_hash     VARCHAR(255) NOT NULL UNIQUE,
```
Comment (Sidarthapati) [inline app/src/main/resources/db/migration/V5__create_api_keys.sql:3]: How does this hash work? I assume we are storing the api key value itself in id, then why are we storing the hash value? Not sure about this, if you could clarify on the process.

File: app/src/main/resources/db/migration/V4__create_users.sql [hunk 2]
```diff
@@ -0,0 +1,15 @@
+CREATE TABLE users (
+    id                   UUID PRIMARY KEY DEFAULT gen_random_uuid(),
+    email                VARCHAR(255) NOT NULL UNIQUE,
+    password_hash        VARCHAR(255) NOT NULL,
```
Comment (Sidarthapati) [inline app/src/main/resources/db/migration/V4__create_users.sql:4]: Same as below.

File: app/src/main/resources/db/migration/V6__create_role_audit_logs.sql [hunk 3]
```diff
@@ -0,0 +1,16 @@
+-- actor_id and target_user_id are plain UUIDs with no FK — audit trail must survive user deletion
+CREATE TABLE role_audit_logs (
+    id             UUID PRIMARY KEY DEFAULT gen_random_uuid(),
+    actor_id       UUID NOT NULL,
```
Comment (Sidarthapati) [inline app/src/main/resources/db/migration/V6__create_role_audit_logs.sql:4]: Just confirming, actor is the guy who is changing the role or making a changes in roles?

File: app/src/main/resources/db/migration/V4__create_users.sql (hunk 2 above)
Comment (BanerjeeAgniva) [reply app/src/main/resources/db/migration/V4__create_users.sql:4]: id is just the internal DB UUID, completely separate from credentials.

password_hash stores BCrypt of the raw password:

Registration / password set:
  raw password: "MySecret123!"
       │
       ▼
  BCrypt("MySecret123!") → "$2a$12$eImiTXuWVxfM37uY4JANjQ..."
       │
       ▼
  password_hash column stores "$2a$12$eImiTXuWVxfM37uY4JANjQ..."
  raw password is discarded immediately

On login:
Incoming: email + "MySecret123!"
       │
       ▼
Fetch user row by email → get stored password_hash
       │
       ▼
BCrypt.verify("MySecret123!", "$2a$12$...") → true/false

One key difference from API keys — BCrypt is specifically designed for passwords:
- It's deliberately slow (work factor $12$ = 2¹² iterations), making brute-force expensive
- It embeds a random salt inside the hash string itself, so two users with the same password get different hashes
- API keys use a faster hash (SHA-256) because the key is already a long random string — no need for the slow+salt treatment

So the users table never holds a recoverable password. If a user forgets theirs, the flow is always reset (generate a new one), never retrieve.

File: app/src/main/resources/db/migration/V6__create_role_audit_logs.sql (hunk 3 above)
Comment (BanerjeeAgniva) [reply app/src/main/resources/db/migration/V6__create_role_audit_logs.sql:4]: Yes — actor is the one making the change, target is the one being changed.

---
Scenario: Admin (Sid) promotes a Delivery Associate (Ravi) to Supervisor.

Actor  = Sid  
Target = Ravi   (DELIVERY_ASSOCIATE → SUPERVISOR)

---
What gets written to role_audit_logs:

id             = 550e8400-e29b-41d4-a716-446655440000   ← new UUID for this log row
actor_id       = 9f1b2c3d-...   ← Sid's users.id
target_user_id = 4a5b6c7d-...   ←Sid's users.id
action         = 'ROLE_CHANGE'
previous_role  = 'DELIVERY_ASSOCIATE'
new_role       = 'SUPERVISOR'
city_id        = 'MUM'          ← scoped to Mumbai; Sid
reason         = 'Promoted due to 6-month performance review'
created_at     = 2026-05-13 09:32:00
updated_at     = 2026-05-13 09:32:00

---
Why actor_id and target_user_id are plain UUIDs with no FK:

If Sid's account is later deactivated or deleted from the users table, this audit row must still exist and still be readable — it's legal/compliance history. A FK constraint would either block the deletion of Sid's user row, or cascade-delete this record. Neither is acceptable for an audit trail. Plain UUIDs let the log survive independently.

action can also be 'DEACTIVATION', 'ACTIVATION', 'CUSTOM_ROLE_ASSIGNED' etc. — not just role changes — which is why it's a free VARCHAR(50) and not a boolean flag.

**Review by Sidarthapati — APPROVED** [review]
LGTM

**Review by Satvik2810 — APPROVED** [review]
